### PR TITLE
Use automatic runtime

### LINKED ISSUE #12: What about use the automatic JSX transform
Not sure why we don't use the [`automatic` JSX transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#manual-babel-setup), so that don't need to add `import {h, Fragment} from 'preact'` at every component that needs JSX or Fragment.

From
https://github.com/preactjs/jest-preset-preact/blob/d0decd11e3248baaf9558db16e3a6a97975f48cc/src/babel.js#L15-L21
to
```js
[
  '@babel/plugin-transform-react-jsx', {
    "runtime": "automatic"
    },
]
```

### PR DESCRIPTION
**Pull request by kentr**
Uses the `runtime: "automatic"` setting for
`@babel/plugin-transform-react-jsx` so that
`h` is automatically imported for `JSX`
transpilation.

Is supposed to also automatically import
`Fragment`, but in brief testing on existing
project, does not do so.

Fixes #12

### CODE REVIEW

File: src/babel.js [hunk 1]
```diff
@@ -15,8 +15,7 @@ module.exports = babelJest.default.createTransformer({
 		[
 			'@babel/plugin-transform-react-jsx',
 			{
-				pragma: 'h',
-				pragmaFrag: 'Fragment',
+				runtime: 'automatic'
```
Comment (rschristian) [inline src/babel.js]:
```suggestion
				runtime: 'automatic',
				importSource: 'preact'
```

This would not work without setting the import source.

**Comment by rschristian** [comment]
> so that h is automatically imported for JSX transpilation.

Not quite, no; what it does is add an import for `preact/jsx-runtime` which comes with a bit of overhead compared to `preact.h`.

We'd need a way to customize this behavior, as there are reasons to avoid using the runtime transform.

**Comment by kentr** [comment]
> Not quite, no; what it does is add an import for preact/jsx-runtime which comes with a bit of overhead compared to preact.h.

Ok. You probably understand this better than I do.

I based that statement on the [plugin docs](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx), which imply that by default only the specific functions are imported (emphasis mine):  _With this runtime enabled, the **functions that JSX compiles to** will be imported automatically._

The examples appear to support this, in that they show named imports for specific functions added by the transform.

> We'd need a way to customize this behavior, as there are reasons to avoid using the runtime transform.

According to that page, the behavior [can be customized per file with comments at the top](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-automatic-runtime-import). Does that satisfy the customization needs?

**Comment by rschristian** [comment]
> I based that statement on the [plugin docs](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx), which imply that by default only the specific functions are imported (emphasis mine): _With this runtime enabled, the **functions that JSX compiles to** will be imported automatically._

Sort of.

The React equivalent of Preact's `h()` is `createElement()`, but rather notably, that's not what the Babel example is showing. It's instead using a separate package export that's a bit of an all-in-one handler for JSX. This has the great effect of ensuring your JSX and Fragment pragmas are always lined up (as experienced here) but is a bit different from plain `h()` (or `createElement()`).

You can take a peek at Preact's implementation (if you're curious) here: https://github.com/preactjs/preact/blob/master/jsx-runtime/src/index.js

> According to that page, the behavior [can be customized per file with comments at the top](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-automatic-runtime-import). Does that satisfy the customization needs?

~~I'd say no. Adding a Babel comment at the top of every single test file because of a breaking change isn't a great experience, tbh.~~

Edit: Actually, no, the JSX pragma cannot be set if Babel is using the runtime transform.

**Comment by kentr** [comment]
> Actually, no, the JSX pragma cannot be set if Babel is using the runtime transform.

Can any Babel config be set for Jest if using this preset? I didn't see a way to do that, which is why I followed up on this change.

**Comment by kentr** [comment]
I did look into a custom config merge, but that was too involved.

**Comment by kentr** [comment]
If there is already a way to get this behavior in `jest-preset-preact` with a simple project config override, then this change is pointless.

**Comment by rschristian** [comment]
> If there is already a way to get this behavior in jest-preset-preact with a simple project config override, then this change is pointless.

Unfortunately it doesn't look like it: https://github.com/preactjs/jest-preset-preact/blob/d0decd11e3248baaf9558db16e3a6a97975f48cc/src/babel.js#L24-L25

We probably want to enable these to allow users to supply their own Babel configs, overwriting whichever default we set here.

**Comment by kentr** [comment]
> We probably want to enable these to allow users to supply their own Babel configs, overwriting whichever default we set here.

If changing to the automatic runtime is a breaking change, maybe a better approach is just enabling these so that the automatic runtime is opt-in by way of overriding config.

Also, if the `preact-cli` Babel config can be picked up automatically by Jest, would that take care the issue of importing `h` and `Fragment`? For me, the impetus behind this PR was so that JSX will work in test runs w/o me having to add explicit imports.

**Comment by rschristian** [comment]
> If changing to the automatic runtime is a breaking change, maybe a better approach is just enabling these so that the automatic runtime is opt-in by way of overriding config.

Yup, agreed.

> Also, if the preact-cli Babel config can be picked up automatically by Jest, would that take care the issue of importing h and Fragment?

[Preact-CLI doesn't actually use the runtime transform](https://github.com/preactjs/preact-cli/blob/a41d498b57599d03aaef67b24f9fd58361099403/packages/cli/src/lib/babel-config.js#L28-L29) at the moment, so no, even if Jest picked up its config (which it wouldn't do by default, you'd have to add your own `babel.config.js` or whatever and import CLI's) it wouldn't do anything.

You _can_ use Preact-CLI without importing `h` and `Fragment`, but that's actually due to [a Webpack plugin injecting them](https://github.com/preactjs/preact-cli/blob/a41d498b57599d03aaef67b24f9fd58361099403/packages/cli/src/lib/webpack/webpack-base-config.js#L329-L332) so they're always available.

Slightly different things, that and the runtime transform.

**Comment by rschristian** [comment]
Shoot, forgot Babel doesn't support having conflicting plugins override. It doesn't look like there's anyway to provide arguments to a Jest preset either.

As this preset is mainly for `preact-cli` use, I think we're going to have to keep it as-is for the moment. I am looking to switch to the automatic runtime in v4, and that might be coming up soon-ish here.

**Comment by kentr** [comment]
> Shoot, forgot Babel doesn't support having conflicting plugins override. It doesn't look like there's anyway to provide arguments to a Jest preset either.

I hit this trying to override babel options with a config file. Config file is read, but programmatic options in the preset always override the config file.

The latest commit appears to work along with a project `babel.config.js` that contains the full `babel` configuration.

It's a little messy b/c it requires repeating the entire `babel` config in `jest-preset-preact`.  

Maybe there is a way to extract the `jest-preset-preact` `babel` config to a file, and then use `extends` in the project-wide config to simply extend the preset config.

```javascript
// babel.config.js in project root.

module.exports = function(api) {
  api.cache(true);
  
  return {
    env: {
      test: {
        presets: [
          [
            '@babel/preset-typescript',
            {
              jsxPragma: 'h',
              jsxPragmaFrag: 'Fragment',
            },
          ],
          '@babel/preset-env',
        ],
        plugins: [
          [
            '@babel/plugin-transform-react-jsx',
            {
              runtime: 'automatic',
            },
          ],
          '@babel/plugin-proposal-class-properties',
        ],
      },
    },
  };
};

```

**Comment by rschristian** [comment]
Yeah that's what I wanted to avoid, especially if this config grows further, as users will essentially need to mirror it into their own projects which ain't great. `extends` would limit overwriting to a single field (`plugins`), so in this case users would still need to add `@babel/plugin-proposal-class-properties` themselves too. Better for sure, but still less than ideal.

Perhaps an env var isn't the worst thing in the world, though I don't exactly love it.

File: src/babel.js [hunk 2]
```diff
@@ -28,6 +26,11 @@ const options = config.files.size
 		],
 		babelrc: false,
 		configFile: false,
+		overrides: config.files.size
```
Comment (kentr) [inline src/babel.js]: What about using `overrides`, set to the project config, if any?

Currently working with a sparse project config file:

```javascript
// babel.config.js in project root.

module.exports = function(api) {
  api.cache(true);

  return {
    env: {
      test: {
        plugins: [
          [
            '@babel/plugin-transform-react-jsx',
            {
              runtime: 'automatic',
            },
          ],
        ],
      },
    },
  };
};
```

**Review by rschristian — APPROVED** [review]
Not totally in love with it, but good enough.

Thanks for implementing this!

File: src/babel.js (hunk 2 above)
Comment (rschristian) [reply src/babel.js]: Do remember you need to also set the `importSource` if you're using the automatic transform.

File: src/babel.js [hunk 3]
```diff
@@ -23,4 +26,10 @@ module.exports = babelJest.default.createTransformer({
...
 	babelrc: false,
 	configFile: false,
-});
+	overrides: config.files.size ? [config.options] : [],
+};
+
+module.exports = {
+	...babelJest.default,
+	createTransformer: () => babelJest.default.createTransformer(options),
+};
```
Comment (rschristian) [inline src/babel.js]: What's the reason behind this change?

Comment (kentr) [reply src/babel.js]: That was accidentally left over from earlier attempts. Please check the latest.

File: src/babel.js (hunk 2 above)
Comment (kentr) [reply src/babel.js]: > Do remember you need to also set the `importSource` if you're using the automatic transform.

Thx. I saw your earlier comment about that. 

I can't tell you why, but in my project tests are now running for me with automatic runtime, even without `importSource`. Adding it to `babel.config.js` doesn't make any difference that I can see.

Comment (rschristian) [reply src/babel.js]: It works as we alias `react/jsx-runtime` to `preact/jsx-runtime`.

However, I'd caution against relying on that as module resolution can be somewhat finicky and the main concern is that you end up consuming `preact/compat/jsx-runtime` (as is a common-ish issue, though not one that's present here at the moment, to be clear). That module is `preact/compat` + `preact/jsx-runtime`, so while it'll usually work, `preact/compat` has side effects that could affect your tests.

I'd recommend being explicit and not relying upon aliases that are internal to the preset. It'll also make your Babel config a lot more portable; you can take that and copy/paste it into other projects without worry if you're not relying on externals like aliases to sort paths out.

Just my 2¢, feel free to ignore.

**Comment by rschristian** [comment]
Thanks for implementing this!
